### PR TITLE
feat(download): add opa_no_oci flag to build without containerd

### PR DESCRIPTION
**Pull request by slonka**
xrel https://github.com/open-policy-agent/opa/issues/6160

### Why the changes in this PR are needed?

I think it's best explained in the issue - https://github.com/open-policy-agent/opa/issues/6160

### What are the changes in this PR?

This pr adds an `opa_no_oci` build flag to exclude containerd dependency.

### Notes to assist PR review:

Look at the issue first - https://github.com/open-policy-agent/opa/issues/6160.

### Further comments:

Nothing.

### CODE REVIEW

File: download/oci_download.go [hunk 1]
```diff
@@ -1,3 +1,5 @@
+//go:build !no_oci
```
Comment (srenatus) [inline download/oci_download.go:1]: Any build flag that only makes sense for OPA should start with `opa_`

File: download/oci_download_unavailable.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+import (
+	"context"
+	"github.com/open-policy-agent/opa/bundle"
+	"github.com/open-policy-agent/opa/plugins/rest"
+)
+
+
+// NewOCI returns a new Downloader that can be started.
+func NewOCI(config Config, client rest.Client, path, storePath string) *OCIDownloader {
```
Comment (srenatus) [inline download/oci_download_unavailable.go:13]: [nit]
```suggestion
func NewOCI(Config, rest.Client, string, string) *OCIDownloader {
```

Perhaps we can also get rid of the comments.

**Review by srenatus — COMMENTED** [review]
Skimmed it. Let's collect more input and see if an ideal solution -- removing containerd while keeping the functionality -- would be another option... 😃

**Comment by ashutosh-narkar** [comment]
As Stephan mentioned, removing the containerd dependency would be ideal. A cursory look at the OCI code makes me think we should be able to do that. @DerGut wdyt?  That of course would be more involved but could be better from a maintenance pov.  

The approach in this PR seems fine. I guess we could have additional such flags in the future to control functionality which does not sound ideal.

**Comment by slonka** [comment]
Hey @DerGut 👋 - would love your input on this by the end of this week. We have a release planed next week so would love to know if this could be merged by then or not.

**Comment by DerGut** [comment]
It definitely seems possible to remove `containerd` as a dependency and only use functionality provided by `oras`.

We can use the [remote.Repository](https://pkg.go.dev/oras.land/oras-go/v2@v2.2.0/registry/remote#Repository) type to pass it to the [oras.Copy](https://pkg.go.dev/oras.land/oras-go/v2@v2.2.0#Copy) call.
Oras also provides some interfaces for authentication, there is the [auth.Client](https://pkg.go.dev/oras.land/oras-go/v2@v2.2.0/registry/remote/auth#Client) that can be set on the repository.

I've played around with it a little bit but couldn't yet get all tests to pass. To integrate our authentication plugins with oras's auth interface, more in-depth understanding would be needed.

@slonka if you need to cut a release by the end of the week, I'd suggest to merge this PR first and iterate on removing `containerd` with a bit more time. I could see myself taking this on. @ashutosh-narkar what do you think?

**Comment by srenatus** [comment]
Let's call it out as temporary in the release notes (or omit it), though -- if the containerd dep was gone, there's no need anymore to disable the functionality; so we could revert this in the future.

**Review by ashutosh-narkar — APPROVED** [review]

**Comment by ashutosh-narkar** [comment]
@slonka can you please squash your commits and we can then get this in. Thanks!

**Comment by slonka** [comment]
@ashutosh-narkar the tests seem to be flaky (different ones fail on the same changes) - can you re-run them?

**Comment by ashutosh-narkar** [comment]
> @ashutosh-narkar the tests seem to be flaky (different ones fail on the same changes) - can you re-run them?

The test failures are not related to your changes. We're looking into it.

**Comment by slonka** [comment]
@ashutosh-narkar - last question: is it safe for us to use `main` branch or should we wait for this to be released?

**Pull request merged**